stem/model.py

```diff
@@ -882,6 +873,251 @@ def __initialise_fields(self):
+
+                for element_id, element in new_mesh.elements.items():
+                    self.gmsh_io.mesh_data["elements"]["LINE_2N"][element_id] = element.node_ids
+
+                mp.mesh = new_mesh
+
+    def __get_maximum_element_id(self) -> int:
+        """
+        Returns the maximum element id within the mesh from the mesh data
+
+        Returns:
+            - int: the maximum element id
+
+        """
+        max_element_id = 0
+        for element_type, mesh_element_info in self.gmsh_io.mesh_data["elements"].items():
+            max_element_id = max(max_element_id, max(mesh_element_info.keys()))
+
+        return int(max_element_id)
+
+    def __get_line_string_end_nodes(self, model_part: ModelPart) -> List[List[int]]:
+        """
+        Finds the nodes at both of the line string ends. A line string end is defined as the node which coincides with
+        a geometry point.
+
+        Args:
+            - model_part (:class:`stem.model_part.ModelPart`): model part from which the spring elements need to be
+                extracted.
+
+        Returns:
+            - List[List[int]]: a list of lists which contains both end nodes for separate each line string.
+
+        """
+
+        # assert mesh and geometry are initialised
+        if model_part.geometry is None:
+            raise ValueError(f"Geometry of model part `{model_part.name}` not yet initialised.")
+
+        if model_part.mesh is None:
+            raise ValueError(f"Mesh of model part `{model_part.name}` not yet initialised.")
+
+        # find end nodes
+        end_nodes = self.__find_end_nodes_of_line_strings(model_part.mesh)
+        # find the node ids corresponding to the geometry points
+        node_ids_at_geometry_points = Utils.find_node_ids_close_to_geometry_nodes(
+            mesh=model_part.mesh, geometry=model_part.geometry, eps=1e-06
+        )
+
+        element_ids_search_space = list(model_part.mesh.elements.keys())
+        node_ids_search_space = list(model_part.mesh.nodes.keys())
+
+        # retrieve the connectivity
+        # node -> elements
+        node_to_elements = self.__map_node_to_elements(model_part.mesh)
+
+        # initialise output list
+        line_node_ids = []
+        # initialise a list for end-point we have already encountered n the clustering algorithm
+        completed_points = []
+        for end_node in end_nodes:
+
+            # only consider the end nodes that are not already in the completed_points list
+            if end_node not in completed_points:
+
+                # remove the end node from the list containing the node ids
+                node_ids_search_space.remove(end_node)
+                first_node_id = None
+
+                # if the point is not the end of the cluster, continue until you find the end of the cluster and include
+                # all the line strings
+                while first_node_id not in end_nodes and len(element_ids_search_space) > 0:
```

Comment:
alternative to while keeping the same algorithm: 

            for _ in range(len(element_ids_search_space)):
                second_node_id = self.__find_next_node_along_line_elements(
                    first_node_id, element_ids_search_space, node_ids_search_space, node_to_elements,
                    model_part.mesh.elements, node_ids_at_geometry_points
                )

                # Add the end nodes to the list (start node, end node)
                line_node_ids.append([first_node_id, second_node_id])

                # Update the next point for the search
                first_node_id = second_node_id

                # Check if the current point is an end point
                if first_node_id in end_nodes:
                    break

            # Add the end point to completed_points to reduce the search space
            completed_points.add(first_node_id)

Reply:
element_ids_search_space is reducing in size within the loop, i dont think looping over this in a for loop is a smart thing to do